### ISSUE TITLE
Ensure METcalcpy installs on NOAA machines

### ISSUE THREAD

**Issue by jprestop** [dtcenter/METcalcpy#111]
Ensure that METcalcpy installs on the following NOAA machines:
- WCOSS (or WCOSS2, if access is available)
- Hera
- Orion

## Describe the Task ##
Julie suggests a working session with Minna and Hank to guide installation and troubleshooting on the above mentioned machines.

### Time Estimate ###
1 to 3 days of work

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
Split 2793541 and 2788881

### Funding Source ###
TBD at start of task.

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by jprestop** [comment]
From Logan from the [20210823 METplus User's Telecon](20210823 METplus User's Telecon):

> Could set up a virtual python environment on Orion, but he’ll be working more on that in the coming weeks. _**A higher priority is getting METplotpy/calcpy setup and running on Hera**_. This won’t be on the radar for WCOSS2 until 2022.

**Comment by jprestop** [comment]
I was unable to begin installing on Hera due to lack of permission to create the top level directory in /contrib.  I submitted a helpdesk ticket for the directory to be created.  Once it is created I will try to install on hera.

In testing on Orion, I received the following message:

Orion-devel-1[34] jprestop$ python
Python 3.7.7 (default, Jun 26 2020, 05:10:03)
 [GCC 7.3.0] :: Intel(R) Corporation on linux
Type "help", "copyright", "credits" or "license" for more information.
Intel(R) Distribution for Python is brought to you by Intel Corporation.

>>> from metcalcpy.util import utils
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/apps/contrib/MET/METcalcpy/METcalcpy-1.1.0-beta2/metcalcpy/util/utils.py", line 19, in <module>
    from statsmodels.tsa.arima.model import ARIMA
ModuleNotFoundError: No module named 'statsmodels'

@bikegeek @fisherhucar 
Is "statsmodels" a Python package that I should request be installed on Orion?

**Comment by hankenstein2** [comment]
tldr; Yes.  

That is a required package that we will need for plotpy and METviewer.

**Comment by jprestop** [comment]
Thank you, @fisherhucar.

**Comment by jprestop** [comment]
I heard back from the Orion helpdesk about installing "statsmodels" for the Python version I am using.  Fortunately, this python module can be installed in user's home directory using pip:

```
module load intelpython/2020.2
python -m pip install --user statsmodels
```

I had no idea this was possible.  I ran that and then reran:

```
>>> from metcalcpy.util import utils
>>> 
```

and it was successful!

Running "utils.round_half_up(2.3)"  (and other numbers) and running "help(utils)" also worked great!  

@bikegeek @fisherhucar Is that all that is needed to verify METcalcpy installation?  Are there any other tests I should run?  Thanks!

**Comment by bikegeek** [comment]
Hi Julie,
I suggest you run pytest on test_agg_stats_and_boot.py in the
METcalcpy/test directory.  I ran it on the feature_v1.1.0-beta2 and develop
branches, where Tatiana removed the bootstrapped and pingouin package
dependencies.

I just ran the following from the test directory:  pytest
test_agg_stats_and_boot.py

You can run some of the other tests if you want.    I've attached the
packages I have in the conda environment I used.  Let me know if you have
any problems.

Regards,
Minna

---------------
Minna Win
Pronouns: she/her
National Center for Atmospheric Research
Developmental Testbed Center
Phone: 303-497-8423
Fax:   303-497-8401
---------------

**Comment by jprestop** [comment]
Thanks, @bikegeek.

I ran:

```
module load contrib
module load intel/2020.2
module load intelpython3/2020.2
module use /apps/contrib/modulefiles
module load metcalcpy/1.1.0-beta2
python -m pip install --user statsmodels
python -m pip install --user pytest
export PATH=/apps/contrib/MET/METcalcpy/METcalcpy-1.1.0-beta2/test:${PATH}
```

Then, I ran:

```
Orion-dtn-2[38] jprestop$ python
Python 3.7.7 (default, Jun 26 2020, 05:10:03) 
[GCC 7.3.0] :: Intel(R) Corporation on linux
Type "help", "copyright", "credits" or "license" for more information.
Intel(R) Distribution for Python is brought to you by Intel Corporation.
>>> import pytest
>>> pytest test_agg_stats_and_boot.py
  File "<stdin>", line 1
    pytest test_agg_stats_and_boot.py
                                 ^
```

I tried giving the full path to the file, but that didn’t work either.

```
>>> pytest /apps/contrib/MET/METcalcpy/METcalcpy-1.1.0-beta2/test/test_agg_stats_and_boot.py
  File "<stdin>", line 1
    pytest /apps/contrib/MET/METcalcpy/METcalcpy-1.1.0-beta2/test/test_agg_stats_and_boot.py
                                                     ^
```

I am pretty ignorant with pytest and don’t know what I am doing wrong.  @bikegeek and @fisherhucar any guidance you can give would be appreciated.

**Comment by jprestop** [comment]
@LoganDawson-NOAA @TaraJensen 

@bikegeek @fisherhucar Could you please take a look at the [comment above](https://github.com/dtcenter/METcalcpy/issues/111#issuecomment-917185351) from 9/10 and provide guidance regarding how to proceed with testing METcalcpy?

**Comment by hankenstein2** [comment]
Lets go with what is possible right now. I think unfortunately we will just have to give you a series of command lines to run to test the various plots we currently have. This actually should give us some decent code coverage since the plots depend on so many other elements.

When/if pytest is able to be installed we can switch back although I think the command line plot list will still be valuable.

I will start working on this now.

**Comment by jprestop** [comment]
@fisherhucar Thank you very much!

**Comment by jprestop** [comment]
@fisherhucar, @bikegeek I received an email from Jason Levit about getting METcalcpy and METplotpy on WCOSS2 for testing.  Is there a way to test METcalcpy directly or is the only testing I will do via METplotpy plots?
Adding @TaraJensen to keep her in the loop.

**Comment by bikegeek** [comment]
If WCOSS has pytest installed, you can run some of the pytest unit tests
(METcalcpy/test/), otherwise, you will need to run some of the METplotpy
plots.  As I recall, these hosts don't have pytest installed. METcalcpy was
initially designed to be imported like some of the other Python libraries.
At the very least, you can verify that it is installed by opening a Python
REPL and importing it:

`import metcalcpy`

and verify that you don't get a "Packages not found" message.

---------------
Minna Win
Pronouns: she/her
National Center for Atmospheric Research
Developmental Testbed Center
Phone: 303-497-8423
Fax:   303-497-8401
---------------

**Comment by jprestop** [comment]
Thanks for the quick reply @bikegeek.  Hopefully, WCOSS2 will have pytest installed.  I may need some guidance going forward, but I may not.  I'll follow up with any questions.

**Comment by hankenstein2** [comment]
Done for several NOAA machines, waiting on WCOSS admins for next round. I think it is useful to keep these particular issues going so we can review what we have asked for.

**Comment by hankenstein2** [comment]
I think this issue is not longer useful/relevant. Since we opened it we have morphed our tracking of tasks elsewhere and discussed and tracked in meetings. If we need to re-open for the next round of releases we can just start a new issue.
